extension/parquet/parquet_reader.cpp

```diff
@@ -239,11 +241,35 @@ ParquetReader::ParquetReader(ClientContext &context, string file_name_, vector<L
 		throw FormatException("Footer length %d is too big for the file of size %d", footer_len, file_size);
 	}
 
-	// read footer into buffer and de-thrift
-	buf.resize(footer_len);
-	fs.Read(*handle, buf.ptr, footer_len, file_size - (footer_len + 8));
+	struct stat status;
+	stat(file_name.c_str(), &status);
+
+	// Check context to see if this file has cached metadata
+	if (context.cache.find(file_name) == context.cache.end()) {
+		// read footer into buffer and de-thrift
+		buf.resize(footer_len);
+		fs.Read(*handle, buf.ptr, footer_len, file_size - (footer_len + 8));
+
+		thrift_unpack((const uint8_t *)buf.ptr, &footer_len, &file_meta_data);
+
+		context.cache[file_name] = make_unique<ParquetFileMetadataCache>(file_meta_data, status.st_mtimespec);
```

Comment:
Here we perform the thrift unpack, followed immediately by a copy into the `ParquetFileMetadataCache`. Instead, directly read into the `ParquetFileMetadataCache` and change `file_meta_data` in the Parquet reader to a pointer. This avoids the (very expensive!) copy here.